test/create-locks/test_lock_shimdma.mlir

```diff
@@ -0,0 +1,98 @@
+//===- test_lock3.mlir -----------------------------------------*- MLIR -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2021 Xilinx Inc.
+//
+//===----------------------------------------------------------------------===//
+
+// RUN: aie-opt --aie-create-locks %s | FileCheck %s
+
+// CHECK-LABELs: module @test_lock_shimdma  {
+// CHECK:   AIE.token(0) {sym_name = "token0"}
+// CHECK:   %0 = AIE.external_buffer 0 : memref<256xi32>
+// CHECK:   %1 = AIE.tile(6, 0)
+// CHECK:   %2 = AIE.lock(%1, 0)
+// CHECK:   %3 = AIE.core(%1)  {
+// CHECK:     AIE.useLock(%2, Acquire, 1)
+// CHECK:     AIE.useLock(%2, Release, 0)
+// CHECK:     AIE.end
+// CHECK:   }
+// CHECK:   %4 = AIE.shimDMA(%1)  {
+// CHECK:     %10 = AIE.dmaStart(S2MM0, ^bb1, ^bb2)
+// CHECK:   ^bb1:  // pred: ^bb0
+// CHECK:     AIE.useLock(%2, Acquire, 0)
+// CHECK:     AIE.dmaBd(<%0 : memref<256xi32>, 0, 256>, 0)
+// CHECK:     AIE.useLock(%2, Release, 1)
+// CHECK:     br ^bb2
+// CHECK:   ^bb2:  // 2 preds: ^bb0, ^bb1
+// CHECK:     AIE.end
+// CHECK:   }
+// CHECK:   %5 = AIE.tile(3, 3)
+// CHECK:   %6 = AIE.lock(%5, 0)
+// CHECK:   %7 = AIE.buffer(%5) : memref<256xi32>
+// CHECK:   %8 = AIE.core(%5)  {
+// CHECK:     AIE.useLock(%6, Acquire, 0)
+// CHECK:     AIE.useLock(%6, Release, 1)
+// CHECK:     AIE.end
+// CHECK:   }
+// CHECK:   %9 = AIE.mem(%5)  {
+// CHECK:     %10 = AIE.dmaStart(MM2S0, ^bb1, ^bb2)
+// CHECK:   ^bb1:  // pred: ^bb0
+// CHECK:     AIE.useLock(%6, Acquire, 1)
+// CHECK:     AIE.dmaBd(<%7 : memref<256xi32>, 0, 256>, 0)
+// CHECK:     AIE.useLock(%6, Release, 0)
+// CHECK:     br ^bb2
+// CHECK:   ^bb2:  // 2 preds: ^bb0, ^bb1
+// CHECK:     AIE.end
+// CHECK:   }
+// CHECK:   AIE.flow(%5, DMA : 0, %1, DMA : 0)
+// CHECK: }
+
+// Generate LockOp in the top-level module
+// Lower UseTokenOp to UseLockOp
+// [Core-Mem] ---> [ShimDMA] (non-neighboring tiles)
+// single producer, single consumer
+module @test_lock_shimdma {
+  AIE.token(0) {sym_name = "token0"}
+  %buf_ext = AIE.external_buffer 0 : memref<256xi32>
+
+  %t60 = AIE.tile(6, 0)
+  %c60 = AIE.core(%t60) {
+    AIE.useToken @token0(Acquire, 2)
+    AIE.useToken @token0(Release, 3)
+    AIE.end
+  }
```

Comment:
Shim tile should not have a CoreOp in theory. The token uses here should be located in a runtime dialect operation. I'm experimenting on a runtime dialect locally and will file a PR when it is more mature.

Reply:
Based on my experiments, a CoreOp associated with a shim tile will not break the aiecc.

Reply:
I don't understand the intention here... can you elaborate?

Reply:
The token uses here should happen on the host CPU sides such that we can avoid data pollution & enable double buffering between DDR and host CPU.

For example in this case, only if the host CPU has acquired token0 with value 2, it should begin to use the data on the corresponding external buffer. Similarly, if value 3 is not released by the host CPU, the shim tile should not launch the DMA. Does this make sense?

Reply:
Got it.. one thing I've pondered is having a host CPU exist in the MLIR, which might be a better place for this.. In the meantime, this is probably good enough..  Maybe you can capture this as a TODO in the file.

Reply:
Added a TODO note here!